### PR TITLE
feat[next]: Add support for IfStmt in ITIR

### PR DESCRIPTION
**Pull request by tehrengruber**
Adds support for stateful if statements in ITIR for ITIR embedded and GTFN backend.

### CODE REVIEW

File: tests/next_tests/integration_tests/multi_feature_tests/iterator_tests/test_if_stmt.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            ),
+            lambda: set_at(
+                as_fieldop(multiply, domain)(2.0, inp),
+                domain,
+                out,
+            ),
+        )
+
+    rng = np.random.default_rng()
+    cond = False
```
Comment (havogt) [inline tests/next_tests/integration_tests/multi_feature_tests/iterator_tests/test_if_stmt.py:68]:
```suggestion
```

File: tests/next_tests/integration_tests/multi_feature_tests/iterator_tests/test_if_stmt.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    inp = gtx.as_field([IDim], rng.normal(size=size))
+    out = gtx.as_field([IDim], np.zeros(size))
+    ref = inp if cond else 2.0 * inp
+
+    run_processor(fencil, program_processor, False, inp, out)
```
Comment (havogt) [inline tests/next_tests/integration_tests/multi_feature_tests/iterator_tests/test_if_stmt.py:73]:
```suggestion
    run_processor(fencil, program_processor, cond, inp, out)
```

File: src/gt4py/next/iterator/embedded.py [hunk 3]
```diff
@@ -1603,6 +1603,14 @@ def set_at(expr: common.Field, domain: common.DomainLike, target: common.Mutable
     operators._tuple_assign_field(target, expr, common.domain(domain))
 
 
+@runtime.if_stmt.register(EMBEDDED)
+def if_stmt(cond: bool, true_branch: Callable[[], None], false_branch: Callable[[], None]) -> None:
```
Comment (havogt) [inline src/gt4py/next/iterator/embedded.py:1607]: Let's document somewhere why we have to represent the branches as callables (or more general how we represent `stmt`s in expression trees.

**Review by havogt — COMMENTED** [review]
not a review, just browsed quickly through the code

**Review by havogt — CHANGES_REQUESTED** [review]
I guess it's fine. Embedded is obviously super ugly...

File: src/gt4py/next/iterator/type_system/inference.py [hunk 4]
```diff
@@ -499,6 +499,11 @@ def visit_Temporary(self, node: itir.Temporary, *, ctx) -> ts.FieldType | ts.Tup
             lambda dtype: ts.FieldType(dims=domain.dims, dtype=dtype), node.dtype
         )
 
+    def visit_IfStmt(self, node: itir.IfStmt, *, ctx) -> None:
+        self.visit(node.cond, ctx=ctx)  # TODO: check is boolean
```
Comment (havogt) [inline src/gt4py/next/iterator/type_system/inference.py:503]: highlighting a TODO

File: src/gt4py/next/iterator/embedded.py (hunk 3 above)
Comment (tehrengruber) [reply src/gt4py/next/iterator/embedded.py:1607]: I've added a docstring here.

| or more general how we represent stmts in expression trees

Makes sense, but I skipped this for now.

File: src/gt4py/next/iterator/type_system/inference.py (hunk 4 above)
Comment (tehrengruber) [reply src/gt4py/next/iterator/type_system/inference.py]: Implemented.

File: src/gt4py/next/iterator/type_system/inference.py [hunk 5]
```diff
@@ -499,6 +499,12 @@ def visit_Temporary(self, node: itir.Temporary, *, ctx) -> ts.FieldType | ts.Tup
             lambda dtype: ts.FieldType(dims=domain.dims, dtype=dtype), node.dtype
         )
 
+    def visit_IfStmt(self, node: itir.IfStmt, *, ctx) -> None:
+        assert node.cond == ts.ScalarType(kind=ts.ScalarKind.BOOL)
```
Comment (havogt) [inline src/gt4py/next/iterator/type_system/inference.py:503]:
```suggestion
        assert isinstance(node.cond.type, ts.ScalarType(kind=ts.ScalarKind.BOOL))
```
I guess

**Review by havogt — APPROVED** [review]

**Pull request merged**